### PR TITLE
Fix broken LCTs reported in HLT_L1SingleMu25 [11_2_X]

### PR DESCRIPTION
**Pull request by dildick**
#### PR description:

An inefficiency was reported in HLT_L1SingleMu25 in CMSSW_11_2_0_pre2, which was not present in CMSSW_11_2_0_pre1 (https://its.cern.ch/jira/browse/CMSLITDPG-903). Investigation showed only ME1/2 LCTs (oddly enough) with invalid pattern numbers. This was fixed with https://github.com/cms-sw/cmssw/commit/9354dff643d672ed1d864559d757a53082355968. 

Further investigation revealed that the CLCT processor readout function was not checking if CLCTs were valid, and that GEM-CSC motherboards do not check if incoming pads are valid. The code was also missing quality control. I added a number of `checkValid` functions that check the properties for ALCT/CLCT/LCT at various stages of the algorithm. LogErrors are printed in case values go out of bounds.

#### PR validation:

Tested on 
- `/RelValZMM_14/CMSSW_11_2_0_pre2-112X_mcRun3_2021_realistic_v2-v1/GEN-SIM-DIGI-RAW`
- `/RelValSingleMuFlatPt2To100/CMSSW_11_0_0-PU25ns_110X_mcRun4_realistic_v3_2026D49PU200-v1/GEN-SIM-DIGI-RAW`
- `/RelValSingleMuFlatPt2To100/CMSSW_11_0_0-110X_mcRun4_realistic_v2_2026D49noPU-v1/GEN-SIM-DIGI-RAW`
I did not see LogErrors of invalid stubs.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Needs a backport to 11_1_X for the HLT TDR. https://github.com/cms-sw/cmssw/pull/30914

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by dildick** [comment]
I noticed that ```if (lct.getKeyWG() > max_wire)``` in ```checkValid``` should be tightened to ```if (lct.getKeyWG() >= max_wire)```

**Comment by civanch** [comment]
+1

**Comment by dildick** [comment]
@rekovic A few more commits are needed in this PR and the 11_1_X PR, following investigations on HLT TDR relval samples. I'm also trying to understand a memory problem in 11_1_X, which may also be present in 11_2_X.

File: L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc [hunk 1]
```diff
@@ -1082,6 +1084,111 @@ void CSCCathodeLCTProcessor::dumpDigis(
...
+  }
+
+  // CLCT quality must be valid
+  // CLCTs require at least 4 layers hit
+  // Run-3: ME1/1 CLCTs require only 3 layers
+  // Run-4: ME2/1 CLCTs require only 3 layers
+  if (clct.getQuality() < nplanes_hit_pattern or clct.getQuality() > max_quality) {
+    edm::LogError("CSCCathodeLCTProcessor")
+        << "CSCLCTDigi with invalid quality: " << clct.getQuality() << "; allowed [0," << max_quality << "]";
+    ;
```
Comment (fwyzard) [inline L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc:1115]: extra `;` ?

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by dildick** [comment]
Blast! The 11_2_X version segfaults in the CSCAnodeLCTProcessor. In a test run on /store/relval/CMSSW_11_2_0_pre2/RelValZMM_14/GEN-SIM-DIGI-RAW/112X_mcRun3_2021_realistic_v2-v1/10000/6E16803D-6000-974D\
-8A32-27D7ED12BD5C.root, it happens in Run 1, Event 10057, LumiSection 101.

**Comment by dildick** [comment]
Found the issue. Temporarily reverting `MAX_ALCTS_PER_PROCESSOR = 2` causes a memory issue in https://github.com/cms-sw/cmssw/blob/CMSSW_11_2_X/L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc#L350. Unlike the CLCT processor, this block 
```
      while (start_bx < stop_bx) {
        if (preTrigger(i_wire, start_bx)) {
          if (infoV > 2)
            showPatterns(i_wire);
          if (patternDetection(i_wire, hits_in_patterns)) {
```
is not bound by `MAX_ALCTS_PER_PROCESSOR`. When `MAX_ALCTS_PER_PROCESSOR` is set to 2, `ALCTContainer_` may not have enough space to store all temporary ALCTs, before reading out the actual selected ALCTs. A posssible solution is to change the 2D array `  CSCALCTDigi ALCTContainer_[CSCConstants::MAX_ALCT_TBINS][CSCConstants::MAX_ALCTS_PER_PROCESSOR]` to 
`std::vector<CSCALCTDigi> ALCTContainer_[CSCConstants::MAX_ALCT_TBINS]`, or even better a 2D vector with 1st dimension `CSCConstants::MAX_ALCT_TBINS` and 2nd dimension initially 0.

**Comment by fwyzard** [comment]
@dildick if I read through the code correctly, `ALCTContainer_` is
1. cleared here: https://github.com/cms-sw/cmssw/blob/5d182778be59af8a26000b207fdafcc22cc89e2e/L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc#L196-L198
2. filled here: https://github.com/cms-sw/cmssw/blob/5d182778be59af8a26000b207fdafcc22cc89e2e/L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc#L330-L335 and/or here: https://github.com/cms-sw/cmssw/blob/5d182778be59af8a26000b207fdafcc22cc89e2e/L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc#L346-L351
3. modified here: https://github.com/cms-sw/cmssw/blob/5d182778be59af8a26000b207fdafcc22cc89e2e/L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc#L962-L964
4. read here: https://github.com/cms-sw/cmssw/blob/5d182778be59af8a26000b207fdafcc22cc89e2e/L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc#L1333-L1337


So
  - if one limits the filling in 2. to `MAX_ALCTS_PER_PROCESSOR`, the memory corruption should be solved - but there is no guarantee that the best ALCTs are used later in 4.
  - if one makes `ALCTContainer_` an array of vectors to temporarily keep more candidates, one must also modify the logic in 3. and 4. to pick the best ones ?

**Comment by dildick** [comment]
Exactly. A draft would look like this https://github.com/cms-sw/cmssw/compare/master...dildick:from-CMSSW_11_2_X_2020-07-23-1100-check-valid-lcts-v4?expand=1#diff-ea6ce032380d7ab09800ef403d701649

**Comment by fwyzard** [comment]
@cmsbuild, please test

File: L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc [hunk 2]
```diff
@@ -1082,6 +1077,111 @@ void CSCCathodeLCTProcessor::dumpDigis(
...
```
Comment (fwyzard) [inline L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc:1108]:
```suggestion
```

**Comment by fwyzard** [comment]
urgent

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by civanch** [comment]
+1

**Comment by silviodonato** [comment]
merge

**Pull request merged**